public/static/app/perf-bb/PerfBBOutliersCtrl.js

```diff
@@ -0,0 +1,269 @@
+        detected: STITCH_CONFIG.PERF.COLL_OUTLIERS,
+        suspicious: STITCH_CONFIG.PERF.COLL_OUTLIERS,
+        muted: STITCH_CONFIG.PERF.COLL_MUTE_OUTLIERS,
+        marked: STITCH_CONFIG.PERF.COLL_MARKED_OUTLIERS,
+    };
+
+    vm.lookBackDays = vm.state.lookBackDays;
+
+    vm.applyFiltering = function() {
+        // Say the form the state is 'pristine'
+        $scope.form.$setPristine();
+        // Update controller state
+        vm.state.lookBackDays = +vm.lookBackDays;
+        loadData();
+    };
+
+    // Enhances filtering state with some contextual meta data
+    // This data is required by expression compiler
+    function getFilteringContext(state) {
+        return _.reduce(state.filtering(), (m, v, k) => {
+            if (!getCol) {
+                return m;
+            }
+            const col = getCol(k);
+            if (!col) return m;  // Error! Associated col does not found
+
+            return m.concat({
+                field: k,
+                term: v,
+                type: col.colDef.type || 'string',
+            })
+        }, []);
+    }
+
+    // Creates aggregation expression, which could be used by Stitch
+    // for given `state`
+    function getAggChain(state) {
+        let chain = [];
+
+        // Check if the state has filtering
+        if (state.filtering) {
+            const filteringChain = MDBQueryAdaptor.compileFiltering(
+                // filtering context enhances state data with important meta data
+                getFilteringContext(state)
+            );
+            // check if filtering query was compiled into something
+            filteringChain && chain.push(filteringChain);
+        }
+
+        if (state.sorting) {
+            const sortingChain = MDBQueryAdaptor.compileSorting(state.sorting);
+            // check if sorting query was compiled into something
+            sortingChain && chain.push(sortingChain);
+        }
+
+        chain.push({$limit: LIMIT});
+        return chain;
+    }
+
+    function hydrateData(docs) {
+        _.each(docs, (doc) => {
+            doc._buildId = EvgUtil.generateBuildId({
+                project: project,
+                revision: doc.revision,
+                buildVariant: doc.variant,
+                createTime: doc.create_time,
+            });
+        });
+    }
+
+    // Required by loadData.
+    let theMostRecentPromise;
+
+    function loadData() {
+        vm.isLoading = true;
+        vm.gridOptions.data = [];
+        theMostRecentPromise = Stitch.use(STITCH_CONFIG.PERF).query((db) => {
+            return db.db(STITCH_CONFIG.PERF.DB_PERF)
+                .collection(modeToCollMap[vm.state.mode])
+                .aggregate(getAggChain(vm.state));
+        });
+
+        const thisPromise = theMostRecentPromise;
+        thisPromise.then((docs) => {
+            // If there is more than one concurrent promise - we want the most recent.
+            if (thisPromise !== theMostRecentPromise) {
```

Comment:
I don't understand this pattern - how come you don't want to run every callback?

Reply:
I copied this from the SignalProcessingCtrl, but I think it is for the case that there are multiple loadData() calls in flight. In that case, we only care about the most recent call and can just drop the earlier calls.